### ISSUE TITLE
代码在attention部分的qkv计算时出了问题？

### ISSUE THREAD

**Issue by hfutly66** [LALBJ/PAI#3]
代码在计算k的时候，出现没有对齐的情况，因为self.k_proj是Linear(in_features=4096, out_features=1024, bias=False)，和后面的view(bsz, q_len, self.num_heads, self.head_dim)里面的4096对不上
key_states = (
        self.k_proj(hidden_states)
        .view(bsz, q_len, self.num_heads, self.head_dim)
        .transpose(1, 2)
    )

**Comment by LALBJ** [comment]
哈喽，感谢关注。

这个问题我猜测可能和你使用的模型有关系。我们在 `llama_new_forward` 方法中的实现，除了注释标注出来的部分，别的地方都是 `LlamaAttention` 中 forward 的实现。而在 LLaMA 中，qkv 的映射矩阵 `in_features` 和 `out_features` 都是 4096 维的。

如果是在别的模型上测试我们的方法，建议从对应模型的 `self-attn.forward` 方法去修改。

**Comment by hfutly66** [comment]
> 哈喽，感谢关注。
> 
> 这个问题我猜测可能和你使用的模型有关系。我们在 `llama_new_forward` 方法中的实现，除了注释标注出来的部分，别的地方都是 `LlamaAttention` 中 forward 的实现。而在 LLaMA 中，qkv 的映射矩阵 `in_features` 和 `out_features` 都是 4096 维的。
> 
> 如果是在别的模型上测试我们的方法，建议从对应模型的 `self-attn.forward` 方法去修改。

谢谢您的回复，但我还有一点不明白的是，为什么我运行一次要两天的时间呢？我把batchsize从1调到32，还把机器弄宕机了。。。

**Comment by LALBJ** [comment]
> > 哈喽，感谢关注。
> > 这个问题我猜测可能和你使用的模型有关系。我们在 `llama_new_forward` 方法中的实现，除了注释标注出来的部分，别的地方都是 `LlamaAttention` 中 forward 的实现。而在 LLaMA 中，qkv 的映射矩阵 `in_features` 和 `out_features` 都是 4096 维的。
> > 如果是在别的模型上测试我们的方法，建议从对应模型的 `self-attn.forward` 方法去修改。
> 
> 谢谢您的回复，但我还有一点不明白的是，为什么我运行一次要两天的时间呢？我把batchsize从1调到32，还把机器弄宕机了。。。结果也没达到预期
> 
> ![image](https://private-user-images.githubusercontent.com/60588672/358546485-64042a25-4e45-4d65-95b6-856e8550746d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwMzYwMDksIm5iZiI6MTcyNDAzNTcwOSwicGF0aCI6Ii82MDU4ODY3Mi8zNTg1NDY0ODUtNjQwNDJhMjUtNGU0NS00ZDY1LTk1YjYtODU2ZTg1NTA3NDZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE5VDAyNDgyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA0ZGJmOWFjMDI0ZTI3NWM5MzA2YTcyMjgwNzI4MzNjZTZkNGE4OGE5MmE4OWQ1ZTUxYmZjYTZjZDM1ZmJhNDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.KObUB0Vam4eYlyJmFVeT8ymNJgWxQJSv7fGn4slRf1Q)

哈喽！关于你的疑问，我依次详细回复下哈：

1. **时间效率问题**：目前支持的两个组件中，`inference intervention` 组件在推理过程中不会引入额外时延。然而，`logits refine` 组件在当前版本中的实现效率确实不尽如人意，其在之前的评估测试中会导致推理时间翻倍。因此，我推测你提到的两天推理时间可能也与原模型本身的推理速度较慢有关。为了更高效地进行测试，建议优先使用 `inference intervention` 组件先测试。

2. **batchsize 设置**：尽管我们的代码允许调整 batchsize，但在推理阶段，为了确保结果的准确性，强烈建议将 batchsize 设置为 1。当 batchsize 不为 1 时，由于 padding 机制的存在，可能会导致结果出现不稳定的浮动。

3. **评估结果**：关于你提到的准确率仅为 50% 的情况，这基本上相当于随机猜测，表明代码中可能存在某些 bug。建议提供更多具体的细节，以便能更准确地定位问题并协助解决问题。

希望这些详细回复能对你有所帮助！如果有其他疑问或需要进一步讨论的地方，请随时告知。
